File: pathplannerlib/src/main/java/com/pathplanner/lib/commands/FollowPathWithEvents.java
package com.pathplanner.lib.commands;

import com.pathplanner.lib.path.PathPlannerPath;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.wpilibj2.command.Command;
import java.util.*;
import java.util.function.Supplier;

/** Command that will run a path following command and trigger event markers along the way */
public class FollowPathWithEvents extends Command {
  private final Command pathFollowingCommand;

  /**
   * Constructs a new FollowPathWithEvents command.
   *
   * @param pathFollowingCommand the command to follow the path
   * @param path the path to follow
   * @param poseSupplier a supplier for the robot's current pose
   * @throws IllegalArgumentException if an event command requires the drive subsystem
   * @deprecated No longer needed, as the path following command will now handle events
   */
  @Deprecated(forRemoval = true)
  public FollowPathWithEvents(
      Command pathFollowingCommand, PathPlannerPath path, Supplier<Pose2d> poseSupplier) {
    this.pathFollowingCommand = pathFollowingCommand;

    m_requirements.addAll(pathFollowingCommand.getRequirements());
  }

  @Override
  public void initialize() {
    pathFollowingCommand.initialize();
  }

  @Override
  public void execute() {
    pathFollowingCommand.execute();
  }

  @Override
  public boolean isFinished() {
    return pathFollowingCommand.isFinished();
  }

  @Override
  public void end(boolean interrupted) {
    pathFollowingCommand.end(interrupted);
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/commands/FollowPathWithEvents.cpp
#include "pathplanner/lib/commands/FollowPathWithEvents.h"

using namespace pathplanner;

FollowPathWithEvents::FollowPathWithEvents(
		std::unique_ptr<frc2::Command> &&pathFollowingCommand,
		std::shared_ptr<PathPlannerPath> path,
		std::function<frc::Pose2d()> poseSupplier) : m_pathFollowingCommand(
		std::move(pathFollowingCommand)), m_path(path), m_poseSupplier(
		poseSupplier), m_isFinished(false) {
	AddRequirements(m_pathFollowingCommand->GetRequirements());
	for (EventMarker &marker : m_path->getEventMarkers()) {
		auto reqs = marker.getCommand()->GetRequirements();

		if (!frc2::RequirementsDisjoint(m_pathFollowingCommand.get(),
				marker.getCommand().get())) {
			throw FRC_MakeError(frc::err::CommandIllegalUse,
					"Events that are triggered during path following cannot require the drive subsystem");
		}

		AddRequirements(reqs);
	}
}

void FollowPathWithEvents::Initialize() {
	m_isFinished = false;

	m_currentCommands.clear();

	frc::Pose2d currentPose = m_poseSupplier();
	for (EventMarker &marker : m_path->getEventMarkers()) {
		marker.reset(currentPose);
	}

	m_markers.clear();
	for (EventMarker &marker : m_path->getEventMarkers()) {
		m_markers.emplace_back(marker, false);
	}

	m_pathFollowingCommand->Initialize();
}

void FollowPathWithEvents::Execute() {
	m_pathFollowingCommand->Execute();
	if (m_pathFollowingCommand->IsFinished()) {
		m_pathFollowingCommand->End(false);
		m_isFinished = true;
	}

	for (std::pair<std::shared_ptr<frc2::Command>, bool> &runningCommand : m_currentCommands) {
		if (!runningCommand.second) {
			continue;
		}

		runningCommand.first->Execute();
		if (runningCommand.first->IsFinished()) {
			runningCommand.first->End(false);
			runningCommand.second = false;
		}
	}

	frc::Pose2d currentPose = m_poseSupplier();
	for (std::pair<EventMarker, bool> &marker : m_markers) {
		if (!marker.second) {
			if (marker.first.shouldTrigger(currentPose)) {
				marker.second = true;

				for (std::pair<std::shared_ptr<frc2::Command>, bool> &runningCommand : m_currentCommands) {
					if (!runningCommand.second) {
						continue;
					}

					if (!frc2::RequirementsDisjoint(runningCommand.first.get(),
							marker.first.getCommand().get())) {
						runningCommand.first->End(true);
						runningCommand.second = false;
					}
				}

				marker.first.getCommand()->Initialize();
				m_currentCommands.emplace_back(marker.first.getCommand(), true);
			}
		}
	}
}

bool FollowPathWithEvents::IsFinished() {
	return m_isFinished;
}

void FollowPathWithEvents::End(bool interrupted) {
	if (interrupted) {
		m_pathFollowingCommand->End(true);
	}

	for (std::pair<std::shared_ptr<frc2::Command>, bool> &runningCommand : m_currentCommands) {
		if (runningCommand.second) {
			runningCommand.first->End(true);
		}
	}
}
